### ISSUE TITLE
[T06] CelebA adapter, dataset registry and dataset commands

### ISSUE THREAD

**Issue by akhmialeuski** [akhmialeuski/fly-connectome-lab#7]
## Goal

Read a locally downloaded CelebA, register its location, validate it and show its contents. The dataset is never downloaded by flystate: the maintainer downloads the official files and registers the directory.

## Depends on

- T01

## Files to create

```text
src/flystate/datasets/__init__.py
src/flystate/datasets/registry.py
src/flystate/datasets/celeba.py
src/flystate/cli/dataset.py            # commands: register, validate, info, inspect
tests/synthetic.py                     # add make_synthetic_celeba()
tests/unit/test_celeba.py
tests/unit/test_registry.py
```

Add runtime dependency `pillow`. Add fixture `synthetic_celeba_dir` to `tests/conftest.py`.

## Expected CelebA files

The adapter searches the registered root recursively (first match by name, sorted paths) for:

| Name | Kind | Format |
| --- | --- | --- |
| `img_align_celeba` | directory | JPEG files `000001.jpg` ... `202599.jpg`, 178 × 218 RGB |
| `identity_CelebA.txt` | file | one line per image: `000001.jpg 2880` |
| `list_landmarks_align_celeba.txt` | file | line 1: image count; line 2: header `lefteye_x lefteye_y righteye_x righteye_y nose_x nose_y leftmouth_x leftmouth_y rightmouth_x rightmouth_y`; then `000001.jpg 69 109 106 113 77 142 73 152 108 154` with any amount of whitespace |
| `list_eval_partition.txt` | file | one line per image: `000001.jpg 0` (0 train, 1 val, 2 test) |

This covers both the original Google Drive layout (`Img/`, `Anno/`, `Eval/`) and the torchvision layout (`celeba/` with all files side by side). Kaggle CSV mirrors do not contain identities and are not supported; the error message says so.

## API

```python
@dataclass(frozen=True)
class CelebAFiles:
    images_dir: Path
    identity_file: Path
    landmarks_file: Path
    partition_file: Path

@dataclass(frozen=True)
class CelebARecord:
    filename: str               # "000123.jpg"
    identity: int
    landmarks: tuple[float, ...]  # 10 values in the order of the header
    partition: int              # 0, 1, 2

class CelebAAdapter:
    def __init__(self, root: Path, expected: tuple[int, int] | None = (202_599, 10_177)) -> None: ...
    def locate(self) -> CelebAFiles: ...                      # DatasetError if something is missing
    def records(self) -> list[CelebARecord]: ...              # sorted by filename, parsed once and cached
    def validate(self, full: bool = False) -> dict[str, object]: ...
    def load_image(self, filename: str) -> np.ndarray: ...    # (218, 178, 3) uint8 RGB via Pillow
    def annotation_fingerprint(self) -> str: ...              # sha256_obj of the sha256 of the three annotation files

def sample_id_for(filename: str) -> str: ...                  # "000123.jpg" -> "celeba-000123"
```

- `validate(full=False)`: annotation files parse; the three files list the same set of filenames; number of images and identities equal `expected` (skipped when `expected is None`); the image directory exists. `full=True` additionally checks that every listed JPEG exists (compare sets from one `os.scandir`). Returns `{"ok": bool, "images": int, "identities": int, "problems": [str, ...], "annotation_fingerprint": str}`.

## Registry (`src/flystate/datasets/registry.py`)

File `paths.datasets / "registry.json"`:

```json
{"celeba": {"kind": "celeba", "path": "/abs/path", "registered_utc": "...",
            "validation": {"ok": true, "full": false, "checked_utc": "...", "annotation_fingerprint": "..."}}}
```

Functions `register(paths, name, root)`, `get(paths, name)`, `set_validation(paths, name, report, full)`, `require_validated(paths, name) -> Path` (raises `DatasetError` with the hint `flystate dataset validate celeba` when the dataset is missing or not validated). Writes are atomic (write a temporary file, then `os.replace`).

## CLI

- `flystate dataset register celeba PATH [--json]`: only the name `celeba` is accepted in POC 1 (exit 2 otherwise). Stores the absolute path and clears the previous validation.
- `flystate dataset validate celeba [--full] [--json]`: runs `validate`, stores the result, exit 1 when not ok.
- `flystate dataset info celeba [--json]`: images, identities, images per identity (min, median, max, number of identities with at least 20 images), partition counts, license line "CelebA: non-commercial research purposes only", registry entry.
- `flystate dataset list [--json]`: all registry entries.
- `flystate dataset inspect celeba FILENAME_OR_SAMPLE_ID [--out PNG] [--json]`: prints the record; with `--out` saves the raw image with the five landmarks drawn as small circles.

## Synthetic CelebA fixture

`make_synthetic_celeba(root: Path, identities: int = 8, images_per_identity: int = 25, seed: int = 0) -> None` writes the torchvision layout: `root/celeba/img_align_celeba/*.jpg` (178 × 218, a random background plus two dark ellipses as eyes at the landmark positions), and the three annotation files with correct formats. Landmarks are drawn around `(69, 109), (106, 113), (87, 132), (73, 152), (108, 154)` with a random similarity transform (rotation within ±10°, scale 0.9 to 1.1, shift within ±5 px). Identity ids start at 1; filenames are numbered consecutively from `000001.jpg`; partition values cycle 0, 0, 0, 1, 2.

## Tests

- Parsing of all three files, including the two header lines of the landmark file.
- `locate` finds files in both layouts (create an `Anno/`, `Eval/`, `Img/` variant in a test).
- `validate` with `expected=None` is ok; a missing JPEG is reported only with `full=True`; a filename missing from one annotation file is reported.
- Registry round trip; `require_validated` raises before validation and returns the path after.
- CLI `register`, `validate`, `info --json`, `inspect --out` on the fixture (the CLI accepts `--expected-counts none` as a hidden option for tests).

## Acceptance criteria

Definition of done from `AGENTS.md`. Maintainer check after merge: `flystate dataset register celeba <path>` and `flystate dataset validate celeba --full` succeed on the real data with 202,599 images and 10,177 identities.

## Out of scope

Subset selection, splits, alignment, caching (T07).

**Comment by akhmialeuski** [comment]
## External dataset acquisition and validation evidence

Evidence backfill dated 2026-09-21. Research follow-up: #39.

The official aligned CelebA input was validated: 202,599 images, 10,177 identities, no reported validation problems. The four source files, exact hashes, original acquisition provenance, registration, dataset info, and validation outputs are linked below.

Registration and successful validation do not establish recognition quality or absence of near-duplicate split leakage. Original annotations/photos and image ZIPs remain external. The external-input manifest also records the exact connectome inputs. Availability is controlled by the upstream providers; use the documented source/request procedure and accept only matching hashes. A changed mirror is not an equivalent reproduction.

<details><summary>Immutable evidence files at the archived commit</summary>

- [provenance/acquisition/celeba-info.json](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/celeba-info.json)
- [provenance/acquisition/celeba-registration.json](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/celeba-registration.json)
- [provenance/acquisition/celeba-validation.json](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/celeba-validation.json)
- [provenance/acquisition/provenance.json](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/provenance.json)
- [provenance/acquisition/t16-dataset-validation.json](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/t16-dataset-validation.json)
- [provenance/acquisition/t16-dataset-validation.log](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/t16-dataset-validation.log)
- [provenance/acquisition/verification.log](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/provenance/acquisition/verification.log)

</details>

[Study README and restoration instructions](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/README.md) · [Complete SHA-256 inventory](https://github.com/akhmialeuski/fly-connectome-lab/blob/be17d84978d2a75f87b532db5f6d0ec553bdde73/research/sequential-visual-memory/2026-09-20-celeba-poc1/checksums.sha256). Large payloads require Git LFS; these links are pinned to the archived commit.

**Comment by akhmialeuski** [comment]
### P0 follow-up: cohort integrity and unused-image reserve

Execution: [#40](https://github.com/akhmialeuski/fly-connectome-lab/issues/40). The original split membership is unchanged; these audits do not score predictions.

| Cohort | Train / validation / historical test | Reserved unused photographs | Identities with three reserved photographs | Exact cross-split duplicates | Perceptual review pairs |
| --- | --- | ---: | ---: | ---: | ---: |
| Smoke, 20 identities | 280 / 60 / 60 | 37 | 11 | 0 | 2 |
| Main, 100 identities | 1,400 / 300 / 300 | 224 | 67 | 0 | 8 |

Reserve screening uses source-file/RGB equality and 64-bit grayscale dHash distance <=3 against historical images and previously accepted reserve candidates. Perceptual matches are conservative review candidates, not proven duplicates. Coverage after screening is lower than the preliminary availability-only count (main: 67 rather than 69 identities with three photos). The smoke and main reserves are cohort-specific lists and may overlap; their counts must not be added as independent photographs.

A balanced new three-image confirmation set for all original identities is unavailable. Reserve images remain excluded from all diagnostic fitting/scoring, and the original inspected test is not reused as fresh confirmation. The first campaign remains exploratory. Full per-image signatures, candidate pairs, rejected reserves, accepted reserves, and provenance are retained for archival.

Raw audit comments: [smoke](https://github.com/akhmialeuski/fly-connectome-lab/issues/40#issuecomment-5762860009), [main](https://github.com/akhmialeuski/fly-connectome-lab/issues/40#issuecomment-5762869013).

**Comment by akhmialeuski** [comment]
### Source-data ceiling for a larger-data follow-up

A read-only audit of the installed, validated CelebA identity annotations found:

- 202,599 photographs across 10,177 identities.
- Maximum: **35 photographs for any one identity**.
- 6,348 identities have at least 20 photographs; 3,661 have at least 25; 2,360 have at least 30; only three have 35.
- No identity has 40, 50, or 100 photographs in this source.

Thus substantially increasing independent examples per identity cannot be achieved merely by changing the current sample-count setting. A newly declared cohort can support somewhat more training photographs while retaining separate validation and confirmation photographs, but any such cohort is a new experiment. The existing 14-per-identity training set, historical splits, and reserved photographs remain unchanged.

This availability audit does not select a new cohort or score any predictions. Exact counts are retained in `execution/source-availability.json` for the new study. The proposed nested learning curve on the current cohort remains the first controlled test of the small-data hypothesis; augmentation, if investigated later, must be grouped by original photograph and reported as correlated views rather than new independent examples.

**Comment by akhmialeuski** [comment]
# Fixed-cohort training-size results

All 48 preregistered attempts completed successfully with clean source commit
`c11b847ea3f5ef6704d5b94ff2812517b04bc092`. Twenty identities and the same
60 validation photographs were held fixed. At 2, 4, and 8 photographs per identity,
five nested training-subset draws were fitted; the 14-photograph endpoint is one
shared fit per representation, not five independent replicates. The neural probe
uses the final saved persistent-state observation. Only diagnostic readout
parameters were trained; this study did not train fly synapses or change dynamics.

| Training photos / identity | Pixel control | Encoded-current control | Neural final-state probe |
| --- | --- | --- | --- |
| 2 | 24.33% (16.67–38.33%) | 28.33% (21.67–38.33%) | 6.00% (3.33–10.00%) |
| 4 | 37.33% (25.00–48.33%) | 37.67% (26.67–43.33%) | 3.67% (0.00–6.67%) |
| 8 | 46.00% (40.00–51.67%) | 45.33% (40.00–50.00%) | 4.67% (1.67–8.33%) |
| 14 | 51.67% | 51.67% | 6.67% |

Entries are mean validation accuracy, with the range across training subsets in
parentheses. Uniform-choice reference is 5%. The mean paired change from 2 to 14
photographs is +27.33 percentage points for pixels (95% identity-cluster bootstrap
interval +15.99 to +39.67), +23.33 for encoded currents (+12.00 to +35.00), and
+0.67 for neural activity (-5.68 to +8.67). These exploratory intervals use 2,000
identity-cluster resamples and are conditional on the fixed cohort and declared
subsets; they do not measure uncertainty over new populations or new neural seeds.

More training photographs clearly help the input controls in this cohort. The
measured neural learning curve stays near chance and does not demonstrate a
memory benefit. This does not establish that more data can never help the fly
model. It motivates testing signal preservation and dynamics before expanding
the training set further. Pixel and encoded controls are diagnostic tools, not
replacement vision models or the intended sequential-memory system.

No historical test or reserved photographs were scored. All 48 per-attempt reports
were posted to issue #42 during execution; execution records retain their URLs.
Full parameters, subset memberships, cross-validation details, coefficients,
prediction rows, elapsed time, and execution logs are preserved in the snapshot.

Independent verification reproduced all 7,920 saved prediction rows from the 48
exported coefficient sets. Maximum absolute probability difference was
1.5543122344752192e-15; predicted labels matched exactly. All seven exported arrays
for each full-data endpoint were exactly equal to its prior diagnostic anchor
(scaler, PCA, classes, coefficients, and intercept). This verification required
no refitting. Detailed records are in `execution/replay-verification.json` and
`execution/full-endpoint-parity.json` within the snapshot.
